src/beamlime/applications/handlers.py

```diff
@@ -3,37 +3,59 @@
 import argparse
 import pathlib
 from dataclasses import dataclass
-from typing import Any, NewType
+from typing import NewType
 
 import scipp as sc
+from scippneutron.io.nexus._json_nexus import JSONGroup
 
 from beamlime.logging import BeamlimeLogger
 
 from ..stateless_workflow import StatelessWorkflow, WorkflowResult
-from .base import HandlerInterface, MessageProtocol
+from .base import HandlerInterface
+from .daemons import DetectorDataReceived, RunStart
 
 Events = NewType("Events", list[sc.DataArray])
 
 
+@dataclass
+class DataReady:
+    content: JSONGroup
+
+
 @dataclass
 class WorkflowResultUpdate:
     content: WorkflowResult
 
 
+class DataAssembler(HandlerInterface):
+    """Data assembler receives data and assembles it into a single data structure."""
+
+    def __init__(self) -> None:
+        from ._nexus_helpers import NexusContainer
+
+        self.nexus_container: NexusContainer
+
+    def set_run_start(self, message: RunStart) -> None:
+        self.nexus_container = message.content
+
+    def assemble_detector_data(self, message: DetectorDataReceived) -> DataReady:
+        self.info("Received data from detector %s", message.content['source_name'])
+        self.nexus_container.insert_ev44(message.content)
```

Comment:
Won't this do an in-place update of `self.nexus_container.nexus_dict`? Since we are forwarding this we must avoid modification (see item in #134). Or is this taken care of somehow in `update` , in a way that I cannot see?

Reply:
You're right. But how do we partially update a nested dictionary without making a copy...?
One nexus structure can be up to GB of size in the memory as far as I understood.

Reply:
We only copy the subdicts that get updated. I presume most of the size is in datasets, or subdicts that do not change.

Reply:
Hmm so you're saying we should copy, for example, ``NXeventdata``...?
But those should be somehow inserted into the whole dataset though.

Reply:
I think you need to copy (not deep-copy) all dicts in the path to the event data as well.

Reply:
You mean we make a nested shallow copy and keep the original one somewhere else?

Reply:
Yes, something like that.